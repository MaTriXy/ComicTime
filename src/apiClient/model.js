export class NetworkError {}
export class UnknownError {
  constructor(statusCode) {
    this.statusCode = statusCode;
  }
}
export class NotFound {}
