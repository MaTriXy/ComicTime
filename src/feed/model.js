export class PublicationSummary {
  constructor(
    id,
    url,
    path,
    title,
    lastIssuesTitles,
    image,
    releaseDate,
    publisher
  ) {
    this.id = id;
    this.url = url;
    this.path = path;
    this.title = title;
    this.lastIssuesTitles = lastIssuesTitles;
    this.lastIssuesNumbers = lastIssuesTitles.map(issueTitle =>
      issueTitle.replace(title, "").trim()
    );
    this.image = image;
    this.releaseDate = releaseDate;
    this.publisher = publisher;
  }
}

export class Page {
  constructor(data, page, totalPages) {
    this.data = data;
    this.page = page;
    this.totalPages = totalPages;
  }
}
