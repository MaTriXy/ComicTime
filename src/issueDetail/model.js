export class IssuePage {
  constructor(number, url) {
    this.number = number;
    this.url = url;
  }
}
export class Issue {
  constructor(id, title, numberOfPages, previousIssueId, nextIssueId, pages) {
    this.id = id;
    this.title = title;
    this.numberOfPages = numberOfPages;
    this.previousIssueId = previousIssueId;
    this.nextIssueId = nextIssueId;
    this.pages = pages;
  }
}
