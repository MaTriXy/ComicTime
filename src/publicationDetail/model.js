export class IssueSummary {
  constructor(id, title, releaseDate) {
    this.id = id;
    this.title = title;
    this.releaseDate = releaseDate;
  }
}
export class Publication {
  constructor(
    id,
    image,
    title,
    status,
    summary,
    publisher,
    releaseDate,
    authors,
    issues = []
  ) {
    this.id = id;
    this.image = image;
    this.title = title;
    this.status = status;
    this.summary = summary;
    this.publisher = publisher;
    this.releaseDate = releaseDate;
    this.authors = authors;
    this.issues = issues;
  }
}
