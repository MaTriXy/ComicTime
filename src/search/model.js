export class PublicationSearchResult {
  constructor(id, title) {
    this.id = id;
    this.title = title;
  }
}
