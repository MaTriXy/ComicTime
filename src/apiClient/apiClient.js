import rest from "rest";
import mime from "rest/interceptor/mime";
import log from "loglevel";
import { Left, Right } from "monet";
import { NetworkError, NotFound, UnknownError } from "./model";

const restClient = rest.wrap(mime);

export const baseUrl = "https://allorigins.me";

export function composeResourceUrl(resource) {
  return (
    "/get?url=" + encodeURIComponent(`http://readcomicsonline.ru${resource}`)
  );
}

export function get(resource) {
  return new Promise(resolve => {
    restClient(baseUrl + composeResourceUrl(resource))
      .then(response => {
        processResponse(response, resolve);
      })
      .catch(error => {
        processException(error, resolve);
      });
  });
}

function processResponse(response, resolve) {
  log.debug(response);
  const statusCode = response.status.code;
  if (statusCode === 200) {
    const pageHtml = response.entity.contents;
    resolve(Right(pageHtml));
  } else if (statusCode === 404) {
    resolve(Left(new NotFound()));
  } else {
    const unknownError = new UnknownError(statusCode);
    log.error(unknownError);
    resolve(Left(unknownError));
  }
}

function processException(error, resolve) {
  log.error(error);
  resolve(Left(new NetworkError()));
}
