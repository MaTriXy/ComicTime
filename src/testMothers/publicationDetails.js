import { IssueSummary, Publication } from "../publicationDetail/model";
import { Issue, IssuePage } from "../issueDetail/model";
import { Some } from "monet";

export const anyPublicationId = "the-flash-2016";
export const anyIssueId = "Annual1";

export const anyPublication = new Publication(
  "the-flash-2016",
  "https://comicimages.com/the-flash-1",
  "The Flash",
  Some("Ongoing"),
  "The best comic series in the world",
  Some("DC"),
  Some("January 2016"),
  Some("Peter Gomez"),
  [
    new IssueSummary("1", "the-flash-2016", "The Flash #1", "January 2016"),
    new IssueSummary("2", "the-flash-2016", "The Flash #2", "February 2016")
  ]
);

export const anyIssueSummary = new IssueSummary(
  "1",
  "detective-comic",
  "Detective Comic #1",
  "January 2018"
);

export const anyIssue = new Issue(
  "1",
  "detective-comic",
  "Detective Comic #1",
  2,
  [
    new IssuePage(1, "https://comicpages.com/1"),
    new IssuePage(2, "https://comicpages.com/2")
  ]
);
