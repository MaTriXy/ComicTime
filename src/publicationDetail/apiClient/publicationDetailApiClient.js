import { get } from "../../apiClient/apiClient";
import cheerio from "cheerio";
import { Some, None, Right, Maybe, Left } from "monet";
import { Publication, IssueSummary } from "../model";

export async function getPublication(id) {
  return get(`/comic/${id}`).then(response => {
    return response.flatMap(htmlPage =>
      extractPublicationFromPage(id, htmlPage)
    );
  });
}

function extractPublicationFromPage(id, htmlPage) {
  const loadedPage = cheerio.load(htmlPage);
  const image = loadedPage(".img-responsive")
    .first()
    .attr("src");
  if (Maybe.fromNull(image).isNone()) {
    return Left();
  }
  const title = loadedPage(".listmanga-header")
    .first()
    .text()
    .trim();
  const status = extractStatus(loadedPage);
  const summary = loadedPage(".manga.well p")
    .first()
    .text()
    .trim();
  const infoTableTitles = loadedPage(".dl-horizontal dt");
  const infoTableValues = loadedPage(".dl-horizontal dd");
  const publisher = extractElementFromInfoTable(
    "Type",
    infoTableTitles,
    infoTableValues
  );
  const releaseDate = extractElementFromInfoTable(
    "Date of release",
    infoTableTitles,
    infoTableValues
  );
  const authors = extractElementFromInfoTable(
    "Author(s)",
    infoTableTitles,
    infoTableValues
  ).map(authorLink => {
    const loadedAuthorLink = cheerio.load(authorLink);
    return loadedAuthorLink.text().trim();
  });
  const issues = extractIssues(id, loadedPage);
  return Right(
    new Publication(
      id,
      image,
      title,
      status,
      summary,
      publisher,
      releaseDate,
      authors,
      issues
    )
  );
}

function extractStatus(loadedPage) {
  if (loadedPage(".label-success").length > 0) {
    return Some("Ongoing");
  } else if (loadedPage(".label-danger").length > 0) {
    return Some("Completed");
  } else {
    return None;
  }
}

function extractElementFromInfoTable(
  elementName,
  infoTableTitles,
  infoTableValues
) {
  let result = None();
  infoTableTitles.each((index, elem) => {
    const loadedElement = cheerio(elem);
    if (loadedElement.text().trim() === elementName && !result.isValue) {
      result = Some(
        infoTableValues
          .eq(index)
          .text()
          .trim()
      );
    }
  });
  return result;
}

function extractIssues(publicationId, loadedPage) {
  const issues = [];
  const chaptersReleaseDate = loadedPage(".date-chapter-title-rtl");
  loadedPage(".chapters li h5 a").each((index, elem) => {
    const loadedElem = cheerio(elem);
    const publicationUrl = loadedElem.attr("href");
    const id = publicationUrl.substring(
      publicationUrl.lastIndexOf("/") + 1,
      publicationUrl.length
    );
    const title = loadedElem.html().trim();
    const releaseDate = chaptersReleaseDate
      .eq(index)
      .text()
      .trim();
    issues.push(
      new IssueSummary(
        publicationId + "-" + id,
        publicationId,
        id,
        title,
        releaseDate
      )
    );
  });
  return issues;
}
