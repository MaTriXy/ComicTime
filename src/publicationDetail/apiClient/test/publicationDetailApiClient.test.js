import nock from "nock";
import { baseUrl, composeResourceUrl } from "../../../apiClient/apiClient";
import { NetworkError, UnknownError, NotFound } from "../../../apiClient/model";
import { getPublication } from "../publicationDetailApiClient";
import {
  publicationNotFoundResponse,
  publicationResponse
} from "./resources/publicationDetailApiResponses";
import { IssueSummary } from "../../model";
import { anyPublicationId } from "../../../testMothers/publicationDetails";

describe("Publication detail API Client", () => {
  it("returns an unknown error if the response status code is an error", async () => {
    const errorResponseCode = 500;
    givenThePublicationDetailApiIsNotWorking(
      anyPublicationId,
      errorResponseCode
    );
    const result = await getPublication(anyPublicationId);
    expect(result.left()).toEqual(new UnknownError(errorResponseCode));
  });

  it("returns a network error if there is any not handled exception", async () => {
    givenThereIsNoConnection();
    const result = await getPublication(anyPublicationId);
    expect(result.left()).toEqual(new NetworkError());
  });

  it("returns a publication not found error if the id does not exist", async () => {
    givenThePublicationDoesNotExist();
    const result = await getPublication(anyPublicationId);
    expect(result.left()).toEqual(new NotFound());
  });

  it("returns the extracted publication information composed by the list of issues and their pages", async () => {
    givenTheApiReturnsThePublicationInformation(anyPublicationId);

    const result = await getPublication(anyPublicationId);

    const publication = result.right();
    expect(publication.id).toEqual(anyPublicationId);
    expect(publication.title).toEqual("The Flash (2016-)");
    expect(publication.image).toEqual(
      "https://readcomicsonline.ru/uploads/manga/the-flash-2016/cover/cover_250x350.jpg"
    );
    expect(publication.status.val).toEqual("Ongoing");
    expect(publication.summary).toEqual(
      '"LIGHTNING STRIKES TWICE" Chapter One\n' +
        "\n" +
        "A new storm brews over Central City and disproves the old adage about lightning never, well...you know. Just as Barry begins to feel overwhelmed fighting crime, a new speedster debutsâ€”but just where did this amazing new friend come from?\n" +
        "\n" +
        "FLASH FACT: \"2016 is the 60th anniversary of Barry Allen becoming The Flash, and it's a privilege to be a part of it,\" says writer Joshua Williamson. \"'The Return of Barry Allen' in FLASH #79 [1993] is one of my favorite comics of all time. It's where I became a Flash fan for life.\""
    );
    expect(publication.publisher.val).toEqual("DC Comics");
    expect(publication.releaseDate.val).toEqual("2016");
    expect(publication.authors.val).toEqual("Joshua Williamson");
    expect(publication.issues.length).toEqual(47);
    const firstIssueInTheList = publication.issues[0];
    expect(firstIssueInTheList).toEqual(
      new IssueSummary(
        "the-flash-2016-Annual1",
        "the-flash-2016",
        "Annual1",
        "The Flash (2016-) #Annual 1",
        "31 Jan. 2018"
      )
    );
  });

  it("returns the publication with the issues summary inside", async () => {
    givenTheApiReturnsThePublicationInformation(anyPublicationId);
    const result = await getPublication(anyPublicationId);
    expect(result.right()).toMatchSnapshot();
  });
});

function givenThePublicationDetailApiIsNotWorking(id, errorStatusCode) {
  nock(baseUrl)
    .get(composeResourceUrl(`/comic/${id}`))
    .reply(errorStatusCode);
}

function givenThereIsNoConnection() {
  nock(baseUrl)
    .get("/notHandledPath")
    .reply(1);
}

function givenThePublicationDoesNotExist(id) {
  nock(baseUrl)
    .get(composeResourceUrl(`/comic/${id}`))
    .reply(200, publicationNotFoundResponse);
}

function givenTheApiReturnsThePublicationInformation(id) {
  nock(baseUrl)
    .get(composeResourceUrl(`/comic/${id}`))
    .reply(200, publicationResponse);
}
