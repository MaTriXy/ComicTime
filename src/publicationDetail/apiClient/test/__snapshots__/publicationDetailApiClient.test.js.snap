// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Publication detail API Client returns the publication with the issues summary inside 1`] = `
Object {
  "isRightValue": true,
  "value": Publication {
    "authors": Object {
      "isValue": true,
      "val": "Joshua Williamson",
    },
    "id": "the-flash-2016",
    "image": "http://readcomicsonline.ru/uploads/manga/the-flash-2016/cover/cover_250x350.jpg",
    "issues": Array [
      IssueSummary {
        "id": "Annual1",
        "releaseDate": "31 Jan. 2018",
        "title": "The Flash (2016-) #Annual 1",
      },
      IssueSummary {
        "id": "46",
        "releaseDate": "09 May. 2018",
        "title": "The Flash (2016-) #46",
      },
      IssueSummary {
        "id": "45",
        "releaseDate": "25 Apr. 2018",
        "title": "The Flash (2016-) #45",
      },
      IssueSummary {
        "id": "44",
        "releaseDate": "11 Apr. 2018",
        "title": "The Flash (2016-) #44",
      },
      IssueSummary {
        "id": "43",
        "releaseDate": "28 Mar. 2018",
        "title": "The Flash (2016-) #43",
      },
      IssueSummary {
        "id": "42",
        "releaseDate": "14 Mar. 2018",
        "title": "The Flash (2016-) #42",
      },
      IssueSummary {
        "id": "41",
        "releaseDate": "28 Feb. 2018",
        "title": "The Flash (2016-) #41",
      },
      IssueSummary {
        "id": "40",
        "releaseDate": "14 Feb. 2018",
        "title": "The Flash (2016-) #40",
      },
      IssueSummary {
        "id": "39",
        "releaseDate": "24 Jan. 2018",
        "title": "The Flash (2016-) #39",
      },
      IssueSummary {
        "id": "38",
        "releaseDate": "10 Jan. 2018",
        "title": "The Flash (2016-) #38",
      },
      IssueSummary {
        "id": "37",
        "releaseDate": "27 Dec. 2017",
        "title": "The Flash (2016-) #37",
      },
      IssueSummary {
        "id": "36",
        "releaseDate": "13 Dec. 2017",
        "title": "The Flash (2016-) #36",
      },
      IssueSummary {
        "id": "35",
        "releaseDate": "22 Nov. 2017",
        "title": "The Flash (2016-) #35",
      },
      IssueSummary {
        "id": "34",
        "releaseDate": "08 Nov. 2017",
        "title": "The Flash (2016-) #34",
      },
      IssueSummary {
        "id": "33",
        "releaseDate": "25 Oct. 2017",
        "title": "The Flash (2016-) #33",
      },
      IssueSummary {
        "id": "32",
        "releaseDate": "11 Oct. 2017",
        "title": "The Flash (2016-) #32",
      },
      IssueSummary {
        "id": "31",
        "releaseDate": "27 Sep. 2017",
        "title": "The Flash (2016-) #31",
      },
      IssueSummary {
        "id": "30",
        "releaseDate": "13 Sep. 2017",
        "title": "The Flash (2016-) #30",
      },
      IssueSummary {
        "id": "29",
        "releaseDate": "23 Aug. 2017",
        "title": "The Flash (2016-) #29",
      },
      IssueSummary {
        "id": "28",
        "releaseDate": "09 Aug. 2017",
        "title": "The Flash (2016-) #28",
      },
      IssueSummary {
        "id": "27",
        "releaseDate": "26 Jul. 2017",
        "title": "The Flash (2016-) #27",
      },
      IssueSummary {
        "id": "26",
        "releaseDate": "12 Jul. 2017",
        "title": "The Flash (2016-) #26",
      },
      IssueSummary {
        "id": "25",
        "releaseDate": "28 Jun. 2017",
        "title": "The Flash (2016-) #25",
      },
      IssueSummary {
        "id": "24",
        "releaseDate": "14 Jun. 2017",
        "title": "The Flash (2016-) #24",
      },
      IssueSummary {
        "id": "23",
        "releaseDate": "31 May. 2017",
        "title": "The Flash (2016-) #23",
      },
      IssueSummary {
        "id": "22",
        "releaseDate": "18 May. 2017",
        "title": "The Flash (2016-) #22",
      },
      IssueSummary {
        "id": "21",
        "releaseDate": "26 Apr. 2017",
        "title": "The Flash (2016-) #21",
      },
      IssueSummary {
        "id": "20",
        "releaseDate": "12 Apr. 2017",
        "title": "The Flash (2016-) #20",
      },
      IssueSummary {
        "id": "19",
        "releaseDate": "22 Mar. 2017",
        "title": "The Flash (2016-) #19",
      },
      IssueSummary {
        "id": "18",
        "releaseDate": "08 Mar. 2017",
        "title": "The Flash (2016-) #18",
      },
      IssueSummary {
        "id": "17",
        "releaseDate": "22 Feb. 2017",
        "title": "The Flash (2016-) #17",
      },
      IssueSummary {
        "id": "16",
        "releaseDate": "08 Feb. 2017",
        "title": "The Flash (2016-) #16",
      },
      IssueSummary {
        "id": "15",
        "releaseDate": "25 Jan. 2017",
        "title": "The Flash (2016-) #15",
      },
      IssueSummary {
        "id": "14",
        "releaseDate": "11 Jan. 2017",
        "title": "The Flash (2016-) #14",
      },
      IssueSummary {
        "id": "13",
        "releaseDate": "14 Dec. 2016",
        "title": "The Flash (2016-) #13",
      },
      IssueSummary {
        "id": "12",
        "releaseDate": "09 Dec. 2016",
        "title": "The Flash (2016-) #12",
      },
      IssueSummary {
        "id": "11",
        "releaseDate": "09 Dec. 2016",
        "title": "The Flash (2016-) #11",
      },
      IssueSummary {
        "id": "10",
        "releaseDate": "09 Dec. 2016",
        "title": "The Flash (2016-) #10",
      },
      IssueSummary {
        "id": "9",
        "releaseDate": "09 Dec. 2016",
        "title": "The Flash (2016-) #9",
      },
      IssueSummary {
        "id": "8",
        "releaseDate": "09 Dec. 2016",
        "title": "The Flash (2016-) #8",
      },
      IssueSummary {
        "id": "7",
        "releaseDate": "09 Dec. 2016",
        "title": "The Flash (2016-) #7",
      },
      IssueSummary {
        "id": "6",
        "releaseDate": "09 Dec. 2016",
        "title": "The Flash (2016-) #6",
      },
      IssueSummary {
        "id": "5",
        "releaseDate": "09 Dec. 2016",
        "title": "The Flash (2016-) #5",
      },
      IssueSummary {
        "id": "4",
        "releaseDate": "09 Dec. 2016",
        "title": "The Flash (2016-) #4",
      },
      IssueSummary {
        "id": "3",
        "releaseDate": "09 Dec. 2016",
        "title": "The Flash (2016-) #3",
      },
      IssueSummary {
        "id": "2",
        "releaseDate": "09 Dec. 2016",
        "title": "The Flash (2016-) #2",
      },
      IssueSummary {
        "id": "1",
        "releaseDate": "09 Dec. 2016",
        "title": "The Flash (2016-) #1",
      },
    ],
    "publisher": Object {
      "isValue": true,
      "val": "DC Comics",
    },
    "releaseDate": Object {
      "isValue": true,
      "val": "2016",
    },
    "status": Object {
      "isValue": true,
      "val": "Ongoing",
    },
    "summary": "&quot;LIGHTNING STRIKES TWICE&quot; Chapter One

A new storm brews over Central City and disproves the old adage about lightning never, well...you know. Just as Barry begins to feel overwhelmed fighting crime, a new speedster debuts&#x2014;but just where did this amazing new friend come from?

FLASH FACT: &quot;2016 is the 60th anniversary of Barry Allen becoming The Flash, and it&apos;s a privilege to be a part of it,&quot; says writer Joshua Williamson. &quot;&apos;The Return of Barry Allen&apos; in FLASH #79 [1993] is one of my favorite comics of all time. It&apos;s where I became a Flash fan for life.&quot;",
    "title": "The Flash (2016-)",
  },
}
`;
